signature/cose/envelope.go

```diff
@@ -0,0 +1,575 @@
+		return 0, err
+	}
+	return getSignatureAlgorithmFromKeySpec(keySpec)
+}
+
+// getSignatureAlgorithmFromKeySpec ensures the signing algorithm satisfies
+// algorithm requirements.
+func getSignatureAlgorithmFromKeySpec(keySpec signature.KeySpec) (cose.Algorithm, error) {
+	switch keySpec.Type {
+	case signature.KeyTypeRSA:
+		switch keySpec.Size {
+		case 2048:
+			return cose.AlgorithmPS256, nil
+		case 3072:
+			return cose.AlgorithmPS384, nil
+		case 4096:
+			return cose.AlgorithmPS512, nil
+		default:
+			return 0, &signature.UnsupportedSigningKeyError{Msg: fmt.Sprintf("RSA: key size %d not supported", keySpec.Size)}
+		}
+	case signature.KeyTypeEC:
+		switch keySpec.Size {
+		case 256:
+			return cose.AlgorithmES256, nil
+		case 384:
+			return cose.AlgorithmES384, nil
+		case 521:
+			return cose.AlgorithmES512, nil
+		default:
+			return 0, &signature.UnsupportedSigningKeyError{Msg: fmt.Sprintf("EC: key size %d not supported", keySpec.Size)}
+		}
+	default:
+		return 0, &signature.UnsupportedSigningKeyError{Msg: "key type not supported"}
+	}
+}
+
+// getSigner returns the built-in implementation of cose.Signer from go-cose
+// or a remote signer implementation of cose.Signer.
+func getSigner(signer signature.Signer) (signer, error) {
+	if localSigner, ok := signer.(signature.LocalSigner); ok {
+		return newLocalSigner(localSigner)
+	}
+	return newRemoteSigner(signer)
+}
+
+// generateProtectedHeaders creates Protected Headers of the COSE envelope
+// during Sign process.
+func generateProtectedHeaders(req *signature.SignRequest, protected cose.ProtectedHeader) error {
+	// signingScheme
+	crit := []interface{}{headerLabelSigningScheme}
+	protected[headerLabelSigningScheme] = string(req.SigningScheme)
+
+	// signingTime/authenticSigningTime
+	signingTimeLabel, ok := signingSchemeTimeLabelMap[req.SigningScheme]
+	if !ok {
+		return &signature.InvalidSignRequestError{Msg: "signing scheme: require notary.x509 or notary.x509.signingAuthority"}
+	}
+	protected[signingTimeLabel] = req.SigningTime.Unix()
+	if signingTimeLabel == headerLabelAuthenticSigningTime {
+		crit = append(crit, headerLabelAuthenticSigningTime)
+	}
+
+	// expiry
+	if !req.Expiry.IsZero() {
+		crit = append(crit, headerLabelExpiry)
+		protected[headerLabelExpiry] = req.Expiry.Unix()
+	}
+
+	// extended attributes
+	for _, elm := range req.ExtendedSignedAttributes {
+		if _, ok := protected[elm.Key]; ok {
+			return &signature.InvalidSignRequestError{Msg: fmt.Sprintf("%q already exists in the protected header", elm.Key)}
+		}
+		if elm.Critical {
+			crit = append(crit, elm.Key)
+		}
+		protected[elm.Key] = elm.Value
+	}
+
+	// critical headers
+	protected[cose.HeaderLabelCritical] = crit
+
+	return nil
+}
+
+// generateUnprotectedHeaders creates Unprotected Headers of the COSE envelope
+// during Sign process.
+func generateUnprotectedHeaders(req *signature.SignRequest, signer signer, unprotected cose.UnprotectedHeader) {
+	// signing agent
+	if req.SigningAgent != "" {
```

Comment:
We should add empty string `'   '` check.

Reply:
> We should add empty string `' '` check.

Shiwei had a different opinion on this one, https://github.com/notaryproject/notation-core-go/pull/75#discussion_r978410229. Since req.SigningAgent is not a user input, looks like we don't need to check empty here?